File: ml/nlp.py
# -*- coding: utf-8 -*-
"""Conversión de lenguaje natural (español) a su FBF simbólica.

Es una de las capacidades del módulo de *asistencia inteligente*: el usuario
escribe una oración como::

    "Llueve y hace frío"               (con p=Llueve, q=Hace frío)
    -> (p ∧ q)

Se implementa con **procesamiento de reglas**:
  1. Normalización (minúsculas + sin acentos).
  2. Sustitución de las descripciones declaradas por sus símbolos (p, q, ...).
  3. Detección de conectivos en español:
       "y"  -> ∧      "o"  -> ∨      "no" -> ¬
       "si A entonces B"  -> (A → B)   "A si y solo si B" -> (A ↔ B)
  4. La cadena resultante se valida con el parser de FBF.
"""

from __future__ import annotations

import re
import unicodedata
from typing import Dict, List

from logica.parser import parsear, FBFError

# Conectivos en español y su símbolo lógico.
_CONECTORES = {
    "↔": "↔",
    "y": "∧",
    "o": "∨",
}

# Palabras 'negras' será: si no aparece al inicio se sustituye por ¬.
_ATOMO_RE = re.compile(r"[a-z][a-z0-9]*")
_ASIGNACION_RE = re.compile(r"([A-Za-z0-9_]+)\s*[:=]\s*([^,;\n]+)")


def _sin_acentos(texto: str) -> str:
    """Elimina tildes y normaliza a minúsculas."""
    texto = unicodedata.normalize("NFKD", texto)
    texto = "".join(c for c in texto if not unicodedata.combining(c))
    return texto.lower()


def parsear_asignaciones(texto: str) -> Dict[str, str]:
    """Convierte ``"p: Llueve, q: Hace frío"`` en ``{"p": "llueve", ...}``.

    Los sinónimos separados por ``/`` se guardan todos en la misma clave de
    búsqueda. Devuelve un dict ``{simbolo: descripción}``.
    """
    resultado: Dict[str, str] = {}
    for m in _ASIGNACION_RE.finditer(texto or ""):
        simbolo = m.group(1).strip()
        desc = _sin_acentos(m.group(2).strip()).strip(" .")
        if not _ATOMO_RE.fullmatch(simbolo):
            continue
        resultado[simbolo] = desc
    return resultado


def _mapa_desc_a_simbolo(simbolos: Dict[str, str]) -> List[tuple]:
    """Construye la lista (descripción → símbolo) ordenada de más larga a corta."""
    mapa = []
    for simbolo, desc in simbolos.items():
        desc = _sin_acentos(desc)
        for sinonimo in desc.split("/"):
            sinonimo = sinonimo.strip()
            if sinonimo:
                mapa.append((sinonimo, simbolo))
    mapa.sort(key=lambda x: -len(x[0]))
    return mapa


def _sustituir_descripciones(oracion: str, simbolos: Dict[str, str]) -> str:
    """Reemplaza las descripciones declaradas por el símbolo de su variable."""
    for desc, simbolo in _mapa_desc_a_simbolo(simbolos):
        patron = r"(?<![a-z])" + re.escape(desc) + r"(?![a-z])"
        oracion = re.sub(patron, f" {simbolo} ", oracion)
    return oracion


def _quitar_no_validos(oracion: str) -> str:
    """Borra puntuación residual; conserva letras, dígitos y símbolos lógicos."""
    permitidos = "abcdefghijklmnopqrstuvwxyz0123456789 ()¬∧∨→↔"
    return "".join(c for c in oracion if c in permitidos)


def _buscar_fuera_parens(oracion: str, token: str) -> int:
    """Índice de `token` (símbolo lógico rodeado de espacios) fuera de paréntesis."""
    prof = 0
    n = len(oracion)
    i = 0
    while i < n:
        c = oracion[i]
        if c == "(":
            prof += 1
        elif c == ")":
            prof = max(0, prof - 1)
        elif prof == 0 and oracion.startswith(token, i) \
                and oracion[i - 1].isspace() \
                and (i + len(token) >= n or oracion[i + len(token)].isspace()):
            return i
        i += 1
    return -1


def _dividir(oracion: str, token: str):
    """Divide la oración en (izquierda, derecha) en la primera aparición del token."""
    pos = _buscar_fuera_parens(oracion, token)
    if pos < 0:
        return None
    izq = oracion[:pos].strip()
    der = oracion[pos + len(token):].strip()
    return izq, der


def _balanceada(oracion: str) -> bool:
    """Verifica que los paréntesis estén balanceados."""
    return oracion.count("(") == oracion.count(")") and \
        not re.search(r"\)\s*\(", oracion.replace(" ", ""))


def _es_atomo_valido(pieza: str) -> bool:
    """True si la pieza es una variable proposicional (p, q1, r22, ...)."""
    return bool(_ATOMO_RE.fullmatch(pieza))


def _traducir(oracion: str) -> str:
    """Traduce recursivamente una oración ya normalizada a una cadena FBF."""
    oracion = " ".join(oracion.split())          # normaliza espacios
    if not oracion:
        raise FBFError("No se encontró contenido para traducir.")

    # 1. Paréntesis explícitos: se respetan.
    if oracion.startswith("(") and oracion.endswith(")") and _balanceada(oracion):
        interior = _traducir(oracion[1:-1])
        return f"({interior})"

    # 2. Bicondicional: "p si y solo si q" -> (p ↔ q)
    pieza = _dividir(oracion.strip(" "), "↔")
    if pieza:
        izq, der = pieza
        return f"({_traducir(izq)} ↔ {_traducir(der)})"

    # 3. Condicional: "si A entonces B" -> (A → B)
    m = re.search(r"^si\s+(.+?)\s+entonces\s+(.+)$", oracion)
    if m:
        return f"({_traducir(m.group(1))} → {_traducir(m.group(2))})"

    # 3b. Condicional ya materializado por reglas previas (p.ej. "solo si").
    pieza = _dividir(oracion, "→")
    if pieza:
        izq, der = pieza
        return f"({_traducir(izq)} → {_traducir(der)})"

    # 4. Disyunción: "A o B" -> (A ∨ B)
    pieza = _dividir(oracion, "∨")
    if pieza:
        izq, der = pieza
        return f"({_traducir(izq)} ∨ {_traducir(der)})"

    # 5. Conjunción: "A y B" -> (A ∧ B)
    pieza = _dividir(oracion, "∧")
    if pieza:
        izq, der = pieza
        return f"({_traducir(izq)} ∧ {_traducir(der)})"

    # 6. Negación explícita: "no A" -> (¬ A)
    if oracion.startswith("no "):
        return f"(¬ {_traducir(oracion[3:])})"

    # 7. Caso base: variable simple.
    if _es_atomo_valido(oracion):
        return oracion

    # Si quedan palabras sueltas que no son conectivos, hubo algo sin declarar.
    sobrantes = [t for t in oracion.split() if not _es_atomo_valido(t)]
    raise FBFError(
        "Término(s) no reconocidos en la oración: "
        + ", ".join(sobrantes)
        + ". Decláralos en las asignaciones (ej. p: Llueve)."
    )


def oracion_a_fbf(oracion: str, simbolos: Dict[str, str]) -> str:
    """Convierte una oración en español a su FBF (cadena canónica).

    Raises:
        FBFError: si no se puede interpretar o la FBF resultante es inválida.
    """
    texto = _sin_acentos(oracion or "")
    texto = _sustituir_descripciones(texto, simbolos)

    # 1) Bicondicional: "p si y solo si q" -> "p ↔ q".
    texto = re.sub(r"\bsi y solo si\b", " ↔ ", texto)
    # 2) "p solo si q" («p sólo si q») se lee como condicional didáctico p → q.
    texto = re.sub(r"^(.+?)\s+solo\s+si\s+(.+)$", r"(\1 → \2)", texto)
    # 3) Conectivos binarios en español (palabras completas, con límites).
    texto = re.sub(r"\by\b", " ∧ ", texto)
    texto = re.sub(r"\bo\b", " ∨ ", texto)

    texto = _quitar_no_validos(texto)
    if not texto.strip():
        raise FBFError("La oración quedó vacía tras la normalización.")

    fbf = _traducir(texto.strip(" ."))
    # Validación final con el parser real de FBF.
    parsear(fbf)
    return fbf

File: web/assets/js/nlp.js
/* =========================================================================
   ml/nlp.py → JavaScript
   Traducción de lenguaje natural (español) a FBF simbólica mediante reglas.
   ========================================================================= */
"use strict";

(function () {
  var LP = window.LogicaParser;
  var N = {};

  var ATOMO_RE = /^[a-z][a-z0-9]*$/i;
  var ASIGNACION_RE = /([A-Za-z0-9_]+)\s*[:=]\s*([^,;\n]+)/g;

  function sinAcentos(texto) {
    return String(texto).normalize("NFKD").replace(/[\u0300-\u036f]/g, "").toLowerCase();
  }

  function parsearAsignaciones(texto) {
    var resultado = {};
    if (!texto) return resultado;
    var m;
    ASIGNACION_RE.lastIndex = 0;
    while ((m = ASIGNACION_RE.exec(texto)) !== null) {
      var simbolo = m[1].trim();
      var desc = sinAcentos(m[2].trim()).replace(/[ .]+$/g, "");
      if (!ATOMO_RE.test(simbolo)) continue;
      resultado[simbolo] = desc;
    }
    return resultado;
  }

  function mapaDescASimbolo(simbolos) {
    var mapa = [];
    Object.keys(simbolos).forEach(function (s) {
      var desc = sinAcentos(simbolos[s]);
      desc.split("/").forEach(function (sn) {
        sn = sn.trim();
        if (sn) mapa.push([sn, s]);
      });
    });
    mapa.sort(function (a, b) { return b[0].length - a[0].length; });
    return mapa;
  }

  function sustituirDescripciones(oracion, simbolos) {
    mapaDescASimbolo(simbolos).forEach(function (par) {
      var desc = par[0], simbolo = par[1];
      var re = new RegExp("(?<![a-z])" + desc.replace(/[.*+?^${}()|[\]\\]/g, "\\$&") + "(?![a-z])", "g");
      oracion = oracion.replace(re, " " + simbolo + " ");
    });
    return oracion;
  }

  function quitarNoValidos(oracion) {
    var permitidos = "abcdefghijklmnopqrstuvwxyz0123456789 ()¬∧∨→↔";
    var out = "";
    for (var i = 0; i < oracion.length; i++) {
      if (permitidos.indexOf(oracion[i]) !== -1) out += oracion[i];
    }
    return out;
  }

  function buscarFueraParens(oracion, token) {
    var prof = 0;
    for (var i = 0; i < oracion.length; i++) {
      var c = oracion[i];
      if (c === "(") prof++;
      else if (c === ")") prof = Math.max(0, prof - 1);
      else if (prof === 0 && oracion.startsWith(token, i) &&
               (i === 0 || oracion[i - 1].charAt(0).match(/\s/)) &&
               (i + token.length >= oracion.length || oracion[i + token.length].charAt(0).match(/\s/))) {
        return i;
      }
    }
    return -1;
  }

  function dividir(oracion, token) {
    var pos = buscarFueraParens(oracion, token);
    if (pos < 0) return null;
    return [oracion.slice(0, pos).trim(), oracion.slice(pos + token.length).trim()];
  }

  function balanceada(oracion) {
    if (oracion.split("(").length - 1 !== oracion.split(")").length - 1) return false;
    return !/\)\s*\(/.test(oracion.replace(/ /g, ""));
  }

  function esAtomoValido(pieza) { return ATOMO_RE.test(pieza); }

  function traducir(oracion) {
    oracion = oracion.replace(/\s+/g, " ").trim();
    if (!oracion) throw new LP.FBFError("No se encontró contenido para traducir.");

    // 1. Paréntesis explícitos.
    if (oracion.charAt(0) === "(" && oracion.charAt(oracion.length - 1) === ")" && balanceada(oracion)) {
      return "(" + traducir(oracion.slice(1, -1)) + ")";
    }

    // 2. Bicondicional: "p si y solo si q".
    var pieza = dividir(oracion, "↔");
    if (pieza) return "(" + traducir(pieza[0]) + " ↔ " + traducir(pieza[1]) + ")";

    // 3. Condicional: "si A entonces B".
    var m = oracion.match(/^si\s+(.+?)\s+entonces\s+(.+)$/);
    if (m) return "(" + traducir(m[1]) + " → " + traducir(m[2]) + ")";

    // 3b. Condicional ya materializado.
    pieza = dividir(oracion, "→");
    if (pieza) return "(" + traducir(pieza[0]) + " → " + traducir(pieza[1]) + ")";

    // 4. Disyunción.
    pieza = dividir(oracion, "∨");
    if (pieza) return "(" + traducir(pieza[0]) + " ∨ " + traducir(pieza[1]) + ")";

    // 5. Conjunción.
    pieza = dividir(oracion, "∧");
    if (pieza) return "(" + traducir(pieza[0]) + " ∧ " + traducir(pieza[1]) + ")";

    // 6. Negación: "no A".
    if (oracion.startsWith("no ")) return "(¬ " + traducir(oracion.slice(3)) + ")";

    // 7. Caso base: variable.
    if (esAtomoValido(oracion)) return oracion;

    var sobrantes = oracion.split(/\s+/).filter(function (t) { return !esAtomoValido(t); });
    throw new LP.FBFError(
      "Término(s) no reconocidos en la oración: " + sobrantes.join(", ") +
      ". Decláralos en las asignaciones (ej. p: Llueve)."
    );
  }

  function oracionAFbf(oracion, simbolos) {
    var texto = sinAcentos(oracion || "");
    texto = sustituirDescripciones(texto, simbolos);

    texto = texto.replace(/\bsi y solo si\b/g, " ↔ ");
    texto = texto.replace(/^(.+?)\s+solo\s+si\s+(.+)$/g, "($1 → $2)");
    texto = texto.replace(/\by\b/g, " ∧ ");
    texto = texto.replace(/\bo\b/g, " ∨ ");

    texto = quitarNoValidos(texto).trim();
    if (!texto) throw new LP.FBFError("La oración quedó vacía tras la normalización.");

    var fbf = traducir(texto.replace(/[ .]+$/g, ""));
    LP.parsear(fbf); // validación final con el parser real
    return fbf;
  }

  N.sinAcentos = sinAcentos;
  N.parsearAsignaciones = parsearAsignaciones;
  N.oracionAFbf = oracionAFbf;

  window.Nlp = N;
})();